docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,75 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `throw` &ndash; Requires an object reference on the stack so not directly impacted since boxing is not permitted.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+- Use of any Generic which doesn’t expect a ByRefLike parameter as a Generic parameter.
+
+## Proposal
+
+A new Attribute API will be defined to indicate which Generic parameters are permissible to be of any type&mdash;including ByRefLike types. The Attribute could be used by the compiler to implement a generic non-constraint.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
```

Comment:
Why not `[AttributeUsage(AttributeTargets.GenericParameter)]` and then we don't need to deal with the index separately?

(But if we expect runtime to be reading this, Jan's suggestion to use the from GenericParamAttributes is even better - though it's a file format change. If the runtime won't be reading this, I would be more inclined towards an attribute.)

Reply:
I agree. I am incline to use `GenericParamAttributes`. I'll be updating this part.